pkg/controller/jobs/deployment/deployment_webhook.go

```diff
@@ -0,0 +1,110 @@
+
+	appsv1 "k8s.io/api/apps/v1"
+	apivalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"sigs.k8s.io/kueue/pkg/controller/constants"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+type Webhook struct {
+	client                     client.Client
+	manageJobsWithoutQueueName bool
+}
+
+func SetupWebhook(mgr ctrl.Manager, opts ...jobframework.Option) error {
+	options := jobframework.ProcessOptions(opts...)
+	wh := &Webhook{
+		client:                     mgr.GetClient(),
+		manageJobsWithoutQueueName: options.ManageJobsWithoutQueueName,
+	}
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&appsv1.Deployment{}).
+		WithDefaulter(wh).
+		WithValidator(wh).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/mutate--v1-deployment,mutating=true,failurePolicy=fail,sideEffects=None,groups="",resources=deployments,verbs=create,versions=v1,name=mdeployment.kb.io,admissionReviewVersions=v1
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+
+var _ webhook.CustomDefaulter = &Webhook{}
+
+func (wh *Webhook) Default(ctx context.Context, obj runtime.Object) error {
+	d := FromObject(obj)
+	log := ctrl.LoggerFrom(ctx).WithName("deployment-webhook").WithValues("deployment", klog.KObj(d))
+	log.V(5).Info("Applying defaults")
+
+	if d.Spec.Template.Labels == nil {
+		d.Spec.Template.Labels = make(map[string]string)
+	}
+	d.Spec.Template.Labels[constants.QueueLabel] = d.Labels[constants.QueueLabel]
+
+	return nil
+}
+
+// +kubebuilder:webhook:path=/validate--v1-deployment,mutating=false,failurePolicy=fail,sideEffects=None,groups="",resources=deployments,verbs=create;update,versions=v1,name=vdeployment.kb.io,admissionReviewVersions=v1
+
+var _ webhook.CustomValidator = &Webhook{}
+
+func (wh *Webhook) ValidateCreate(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	return nil, nil
+}
+
+var (
+	deploymentLabelsPath         = field.NewPath("metadata", "labels")
+	deploymentQueueNameLabelPath = deploymentLabelsPath.Key(constants.QueueLabel)
+
+	podSpecLabelsPath         = field.NewPath("spec", "template", "metadata", "labels")
+	podSpecQueueNameLabelPath = podSpecLabelsPath.Key(constants.QueueLabel)
+)
+
+func (wh *Webhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldDeployment := FromObject(oldObj)
+	newDeployment := FromObject(newObj)
+
+	log := ctrl.LoggerFrom(ctx).WithName("deployment-webhook").WithValues("deployment", klog.KObj(newDeployment))
+	log.V(5).Info("Validating update")
+	allErrs := apivalidation.ValidateImmutableField(
+		newDeployment.GetLabels()[constants.QueueLabel],
+		oldDeployment.GetLabels()[constants.QueueLabel],
```

Comment:
I guess that this has a potential null pointer error since the `GetLabels` function will return nil if the Deployment does not have any labels, right?

https://github.com/kubernetes/apimachinery/blob/a8a2284d318b9213fca20b0058e969941ce1c90b/pkg/apis/meta/v1/meta.go#L163

Reply:
This is fine, though https://go.dev/play/p/pP1v9ALY6hh

Reply:
Oh, there are no error even though that is null reference. It seems to be tricky.

Thank you for confirmation.